diff -NaBur -x .git fabm-git-0/compilers/vs2008/fabm-glm.vfproj fabm-git/compilers/vs2008/fabm-glm.vfproj
--- fabm-git-0/compilers/vs2008/fabm-glm.vfproj	2014-09-19 09:58:37.000000000 +0800
+++ fabm-git/compilers/vs2008/fabm-glm.vfproj	2014-09-19 10:16:57.586427346 +0800
@@ -78,10 +78,10 @@
 		<File RelativePath="..\..\src\Makefile"/>
 		<File RelativePath="..\..\src\Rules.make"/></Filter>
 		<Filter Name="Source Files" Filter="f90;for;f;fpp;ftn;def;odl;idl">
-		<Filter Name="config">
-		<File RelativePath="..\..\src\config\fabm_config.F90"/>
-		<File RelativePath="..\..\src\config\fabm_config_types.F90"/>
-		<File RelativePath="..\..\src\config\fabm_yaml.F90"/></Filter>
+		<Filter Name="config">
+		<File RelativePath="..\..\src\config\fabm_config.F90"/>
+		<File RelativePath="..\..\src\config\fabm_config_types.F90"/>
+		<File RelativePath="..\..\src\config\fabm_yaml.F90"/></Filter>
 		<Filter Name="models">
 		<Filter Name="aed">
 		<File RelativePath="..\..\src\models\aed\aed.h"/>
@@ -112,7 +112,8 @@
 		<Filter Name="filter_feeder">
 		<File RelativePath="..\..\src\models\bb\filter_feeder\filter_feeder.F90"/></Filter>
 		<Filter Name="passive">
-		<File RelativePath="..\..\src\models\bb\passive\passive.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\bb\passive\passive.F90"/></Filter>
+		<File RelativePath="..\..\src\models\bb\bb_model_library.F90"/></Filter>
 		<Filter Name="examples">
 		<Filter Name="benthic_predator">
 		<File RelativePath="..\..\src\models\examples\benthic_predator\benthic_predator.F90"/></Filter>
@@ -122,28 +123,32 @@
 		<File RelativePath="..\..\src\models\examples\mean\mean.F90"/></Filter>
 		<Filter Name="npzd">
 		<File RelativePath="..\..\src\models\examples\npzd\det.F90"/>
+		<File RelativePath="..\..\src\models\examples\npzd\examples_npzd.F90"/>
 		<File RelativePath="..\..\src\models\examples\npzd\nut.F90"/>
 		<File RelativePath="..\..\src\models\examples\npzd\phy.F90"/>
 		<File RelativePath="..\..\src\models\examples\npzd\zoo.F90"/></Filter>
 		<Filter Name="npzd_f2003">
-		<File RelativePath="..\..\src\models\examples\npzd_f2003\npzd_f2003.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\examples\npzd_f2003\npzd_f2003.F90"/></Filter>
+		<File RelativePath="..\..\src\models\examples\examples_model_library.F90"/></Filter>
 		<Filter Name="gotm">
 		<Filter Name="ergom">
 		<File RelativePath="..\..\src\models\gotm\ergom\ergom.F90"/></Filter>
 		<Filter Name="fasham">
 		<File RelativePath="..\..\src\models\gotm\fasham\fasham.F90"/></Filter>
 		<Filter Name="npzd">
-		<File RelativePath="..\..\src\models\gotm\npzd\npzd.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\gotm\npzd\npzd.F90"/></Filter>
+		<File RelativePath="..\..\src\models\gotm\gotm_model_library.F90"/></Filter>
 		<Filter Name="hzg">
 		<Filter Name="omexdia_p">
 		<File RelativePath="..\..\src\models\hzg\omexdia_p\omexdia_p.F90"/></Filter></Filter>
 		<Filter Name="iow">
 		<Filter Name="age">
 		<File RelativePath="..\..\src\models\iow\age\iow_age.F90"/></Filter>
-		<Filter Name="ergom">
-		<File RelativePath="..\..\src\models\iow\ergom\ergom_cgt.F90"/></Filter>
+		<Filter Name="ergom">
+		<File RelativePath="..\..\src\models\iow\ergom\ergom_cgt.F90"/></Filter>
 		<Filter Name="spm">
-		<File RelativePath="..\..\src\models\iow\spm\spm.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\iow\spm\spm.F90"/></Filter>
+		<File RelativePath="..\..\src\models\iow\iow_model_library.F90"/></Filter>
 		<Filter Name="Klimacampus">
 		<File RelativePath="..\..\src\models\klimacampus\phy_feedback\phy_feedback.F90"/></Filter>
 		<Filter Name="metu">
@@ -152,18 +157,20 @@
 		<Filter Name="msi">
 		<Filter Name="ergom1">
 		<File RelativePath="..\..\src\models\msi\ergom1\ergom1.F90"/></Filter></Filter>
-		<Filter Name="niva">
-		<Filter Name="oxydep">
-		<File RelativePath="..\..\src\models\niva\oxydep\oxydep.F90"/></Filter>
-		<File RelativePath="..\..\src\models\niva\niva_model_library.F90"/></Filter>
+		<Filter Name="niva">
+		<Filter Name="oxydep">
+		<File RelativePath="..\..\src\models\niva\oxydep\oxydep.F90"/></Filter>
+		<File RelativePath="..\..\src\models\niva\niva_model_library.F90"/></Filter>
 		<Filter Name="pml">
 		<File RelativePath="..\..\src\models\pml\carbonate\carbonate.F90"/>
-		<File RelativePath="..\..\src\models\pml\carbonate\co2_dyn.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\pml\carbonate\co2_dyn.F90"/></Filter></Filter>
 		<File RelativePath="..\..\src\fabm.F90"/>
+		<File RelativePath="..\..\src\fabm_builtin_models.F90"/>
+		<File RelativePath="..\..\src\fabm_coupling.F90"/>
 		<File RelativePath="..\..\src\fabm_driver.F90"/>
 		<File RelativePath="..\..\src\fabm_expressions.F90"/>
 		<File RelativePath="..\..\src\fabm_library.F90"/>
-		<File RelativePath="..\..\src\fabm_particle.F90"/>
+		<File RelativePath="..\..\src\fabm_particle.F90"/>
 		<File RelativePath="..\..\src\fabm_properties.F90"/>
 		<File RelativePath="..\..\src\fabm_standard_variables.F90"/>
 		<File RelativePath="..\..\src\fabm_types.F90"/></Filter></Files>
diff -NaBur -x .git fabm-git-0/compilers/vs2010/fabm-glm.vfproj fabm-git/compilers/vs2010/fabm-glm.vfproj
--- fabm-git-0/compilers/vs2010/fabm-glm.vfproj	2014-09-19 09:58:37.000000000 +0800
+++ fabm-git/compilers/vs2010/fabm-glm.vfproj	2014-09-19 10:16:57.586427346 +0800
@@ -78,12 +78,13 @@
 		<File RelativePath="..\..\src\Makefile"/>
 		<File RelativePath="..\..\src\Rules.make"/></Filter>
 		<Filter Name="Source Files" Filter="f90;for;f;fpp;ftn;def;odl;idl">
-		<Filter Name="config">
-		<File RelativePath="..\..\src\config\fabm_config.F90"/>
-		<File RelativePath="..\..\src\config\fabm_config_types.F90"/>
-		<File RelativePath="..\..\src\config\fabm_yaml.F90"/></Filter>
+		<Filter Name="config">
+		<File RelativePath="..\..\src\config\fabm_config.F90"/>
+		<File RelativePath="..\..\src\config\fabm_config_types.F90"/>
+		<File RelativePath="..\..\src\config\fabm_yaml.F90"/></Filter>
 		<Filter Name="models">
 		<Filter Name="aed">
+		<File RelativePath="..\..\src\models\aed\aed.h"/>
 		<File RelativePath="..\..\src\models\aed\aed_carbon.F90"/>
 		<File RelativePath="..\..\src\models\aed\aed_chlorophylla.F90"/>
 		<File RelativePath="..\..\src\models\aed\aed_core.F90"/>
@@ -111,7 +112,8 @@
 		<Filter Name="filter_feeder">
 		<File RelativePath="..\..\src\models\bb\filter_feeder\filter_feeder.F90"/></Filter>
 		<Filter Name="passive">
-		<File RelativePath="..\..\src\models\bb\passive\passive.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\bb\passive\passive.F90"/></Filter>
+		<File RelativePath="..\..\src\models\bb\bb_model_library.F90"/></Filter>
 		<Filter Name="examples">
 		<Filter Name="benthic_predator">
 		<File RelativePath="..\..\src\models\examples\benthic_predator\benthic_predator.F90"/></Filter>
@@ -121,49 +123,54 @@
 		<File RelativePath="..\..\src\models\examples\mean\mean.F90"/></Filter>
 		<Filter Name="npzd">
 		<File RelativePath="..\..\src\models\examples\npzd\det.F90"/>
+		<File RelativePath="..\..\src\models\examples\npzd\examples_npzd.F90"/>
 		<File RelativePath="..\..\src\models\examples\npzd\nut.F90"/>
 		<File RelativePath="..\..\src\models\examples\npzd\phy.F90"/>
 		<File RelativePath="..\..\src\models\examples\npzd\zoo.F90"/></Filter>
 		<Filter Name="npzd_f2003">
-		<File RelativePath="..\..\src\models\examples\npzd_f2003\npzd_f2003.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\examples\npzd_f2003\npzd_f2003.F90"/></Filter>
+		<File RelativePath="..\..\src\models\examples\examples_model_library.F90"/></Filter>
 		<Filter Name="gotm">
 		<Filter Name="ergom">
 		<File RelativePath="..\..\src\models\gotm\ergom\ergom.F90"/></Filter>
 		<Filter Name="fasham">
 		<File RelativePath="..\..\src\models\gotm\fasham\fasham.F90"/></Filter>
 		<Filter Name="npzd">
-		<File RelativePath="..\..\src\models\gotm\npzd\npzd.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\gotm\npzd\npzd.F90"/></Filter>
+		<File RelativePath="..\..\src\models\gotm\gotm_model_library.F90"/></Filter>
 		<Filter Name="hzg">
 		<Filter Name="omexdia_p">
 		<File RelativePath="..\..\src\models\hzg\omexdia_p\omexdia_p.F90"/></Filter></Filter>
 		<Filter Name="iow">
 		<Filter Name="age">
-		<File RelativePath="..\..\src\models\iow\age\iow_age.F90"/></Filter>
-		<Filter Name="ergom">
-		<File RelativePath="..\..\src\models\iow\ergom\ergom_cgt.F90"/></Filter>
+		<File RelativePath="..\..\src\models\iow\age\iow_age.F90"/></Filter>
+		<Filter Name="ergom">
+		<File RelativePath="..\..\src\models\iow\ergom\ergom_cgt.F90"/></Filter>
 		<Filter Name="spm">
-		<File RelativePath="..\..\src\models\iow\spm\spm.F90"/></Filter></Filter>
-		<Filter Name="klimacampus">
-		<Filter Name="phy_feedback">
-		<File RelativePath="..\..\src\models\klimacampus\phy_feedback\phy_feedback.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\iow\spm\spm.F90"/></Filter>
+		<File RelativePath="..\..\src\models\iow\iow_model_library.F90"/></Filter>
+		<Filter Name="Klimacampus">
+		<File RelativePath="..\..\src\models\klimacampus\phy_feedback\phy_feedback.F90"/></Filter>
 		<Filter Name="metu">
 		<Filter Name="mnemiopsis">
 		<File RelativePath="..\..\src\models\metu\mnemiopsis\mnemiopsis.F90"/></Filter></Filter>
 		<Filter Name="msi">
 		<Filter Name="ergom1">
 		<File RelativePath="..\..\src\models\msi\ergom1\ergom1.F90"/></Filter></Filter>
-		<Filter Name="niva">
-		<Filter Name="oxydep">
-		<File RelativePath="..\..\src\models\niva\oxydep\oxydep.F90"/></Filter>
-		<File RelativePath="..\..\src\models\niva\niva_model_library.F90"/></Filter>
+		<Filter Name="niva">
+		<Filter Name="oxydep">
+		<File RelativePath="..\..\src\models\niva\oxydep\oxydep.F90"/></Filter>
+		<File RelativePath="..\..\src\models\niva\niva_model_library.F90"/></Filter>
 		<Filter Name="pml">
 		<File RelativePath="..\..\src\models\pml\carbonate\carbonate.F90"/>
-		<File RelativePath="..\..\src\models\pml\carbonate\co2_dyn.F90"/></Filter></Filter>
+		<File RelativePath="..\..\src\models\pml\carbonate\co2_dyn.F90"/></Filter></Filter>
 		<File RelativePath="..\..\src\fabm.F90"/>
+		<File RelativePath="..\..\src\fabm_builtin_models.F90"/>
+		<File RelativePath="..\..\src\fabm_coupling.F90"/>
 		<File RelativePath="..\..\src\fabm_driver.F90"/>
 		<File RelativePath="..\..\src\fabm_expressions.F90"/>
 		<File RelativePath="..\..\src\fabm_library.F90"/>
-		<File RelativePath="..\..\src\fabm_particle.F90"/>
+		<File RelativePath="..\..\src\fabm_particle.F90"/>
 		<File RelativePath="..\..\src\fabm_properties.F90"/>
 		<File RelativePath="..\..\src\fabm_standard_variables.F90"/>
 		<File RelativePath="..\..\src\fabm_types.F90"/></Filter></Files>
diff -NaBur -x .git fabm-git-0/src/drivers/glm/fabm_driver.h fabm-git/src/drivers/glm/fabm_driver.h
--- fabm-git-0/src/drivers/glm/fabm_driver.h	2014-09-19 09:58:37.000000000 +0800
+++ fabm-git/src/drivers/glm/fabm_driver.h	2014-09-19 10:16:57.586427346 +0800
@@ -1,11 +1,7 @@
 #define _FABM_DIMENSION_COUNT_ 1
-#define _FABM_HORIZONTAL_IS_SCALAR_
 
 ! Include FABM preprocessor definitions.
 ! This *must* be done after the host-specific variables are defined (above),
 ! because these are used in fabm.h.
 #include "fabm.h"
 
-! Not used by FABM itself; only by GLM-FABM driver glm_fabm.F90
-#define _LOCATION_RANGE_ 1:_LOCATION_
-
